pkg/nsx/services/nsxserviceaccount/cluster.go

```diff
@@ -0,0 +1,289 @@
+			Revision:     &revision1,
+			ResourceType: &antreaClusterResourceType,
+			Certificate:  &cert,
+			VhcPath:      &vpcPath,
+			Tags:         s.buildBasicTags(obj),
+		})
+		if err != nil {
+			return err
+		}
+		s.ClusterControlPlaneStore.Add(ccp)
+		clusterId = *ccp.NodeId
+	}
+
+	// create Secret
+	secretName := obj.Name + SecretSuffix
+	secretNamespace := obj.Namespace
+	if err := s.Client.Create(ctx, &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: secretNamespace,
+			// TODO: Add labels/annotations
+			Labels:      nil,
+			Annotations: nil,
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion:         obj.APIVersion,
+				Kind:               obj.Kind,
+				Name:               obj.Name,
+				UID:                obj.UID,
+				Controller:         nil,
+				BlockOwnerDeletion: nil,
+			}},
+			Finalizers: nil,
+		},
+		Immutable: nil,
+		// TODO: Add NSX CA
+		Data: map[string][]byte{SecretCertName: []byte(cert), SecretKeyName: []byte(key)},
+		Type: "",
+	}); err != nil {
+		return err
+	}
+
+	// update NSXServiceAccountStatus
+	obj.Status.Phase = v1alpha1.NSXServiceAccountPhaseRealized
+	obj.Status.Reason = "Success."
+	obj.Status.NSXManagers = s.NSXConfig.NsxApiManagers
+	obj.Status.ClusterID = clusterId
+	obj.Status.ClusterName = clusterName
+	obj.Status.Secrets = []v1alpha1.NSXSecret{{
+		Name:      secretName,
+		Namespace: secretNamespace,
+	}}
+	obj.Status.VPCPath = vpcPath
+	// TODO: Add proxy
+	return s.Client.Status().Update(ctx, obj)
+}
+
+func (s *NSXServiceAccountService) DeleteNSXServiceAccount(ctx context.Context, namespacedName types.NamespacedName) error {
+	clusterName := s.getClusterName(namespacedName.Namespace, namespacedName.Name)
+	normalizedClusterName := util.NormalizeId(clusterName)
+	// delete Secret
+	secretName := namespacedName.Name + SecretSuffix
+	secretNamespace := namespacedName.Namespace
+	if err := s.Client.Delete(ctx, &v1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: secretNamespace}}); err != nil && !errors.IsNotFound(err) {
+		log.Error(err, "Failed to delete", "secret", secretName, "namespace", secretNamespace)
+		return err
+	}
+
+	// delete ClusterControlPlane
+	cascade := true
+	if err := s.NSXClient.ClusterControlPlanesClient.Delete(siteId, enforcementpointId, normalizedClusterName, &cascade); err != nil {
+		log.Error(err, "Failed to delete", "ClusterControlPlane", normalizedClusterName)
+		return err
+	}
+	s.ClusterControlPlaneStore.Delete(model.ClusterControlPlane{Id: &normalizedClusterName})
+
+	// delete PI
+	if piobj := s.PrincipalIdentityStore.GetByKey(normalizedClusterName); piobj != nil {
+		pi := piobj.(mpmodel.PrincipalIdentity)
+		if err := s.NSXClient.PrincipalIdentitiesClient.Delete(*pi.Id); err != nil {
+			log.Error(err, "Failed to delete", "PrincipalIdentity", *pi.Name)
+			return err
+		}
+		if pi.CertificateId != nil && *pi.CertificateId != "" {
```

Comment:
if there is case: PI is deleted from NSX but not from store yet, but certificate is not deleted. 
will s.NSXClient.PrincipalIdentitiesClient.Delete raise error for a already deleted PI id?

Reply:
No, most DELETE NSXAPIs will return `200 OK` if the `Id` is not present.

Thus we delete the resource directly for CCP and PI.

Reply:
thanks, then this LGTM.